crates/autopilot/src/domain/events/mod.rs

```diff
@@ -0,0 +1,205 @@
+use {
+    crate::{
+        boundary,
+        domain::{self, eth},
+        infra::{
+            self,
+            persistence::{self, transaction::Transaction},
+        },
+    },
+    anyhow::Context,
+    ethrpc::current_block::RangeInclusive,
+    std::sync::Arc,
+};
+
+pub struct Events {
+    persistence: Arc<infra::Persistence>,
+}
+
+/// Error type for the events module.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("persistence error: {0:?}")]
+    Persistence(#[from] persistence::Error),
+    #[error("unexpected event data: {0:?}")]
+    Encoding(anyhow::Error),
+}
+
+pub mod settlement {
+    use super::*;
+
+    /// An order was fully or partially traded.
+    pub struct Trade {
+        pub block_number: u64,
+        /// The index of the event in the block.
+        pub log_index: usize,
+        pub uid: domain::OrderUid,
+        pub sell_amount_including_fee: eth::U256,
+        pub buy_amount: eth::U256,
+        pub fee_amount: eth::U256,
+    }
+
+    /// An order was cancelled on-chain.
+    pub struct Cancellation {
+        pub block_number: u64,
+        /// The index of the event in the block.
+        pub log_index: usize,
+        pub uid: domain::OrderUid,
+    }
+
+    /// A settlement was executed on-chain.
+    pub struct Settlement {
+        pub block_number: u64,
+        /// The index of the event in the block.
+        pub log_index: usize,
+        pub tx_hash: eth::H256,
+        pub solver: eth::Address,
+    }
+
+    /// An order was signed on-chain or signature for an on-chain signed order
+    /// has been revoked.
+    pub struct PreSignature {
+        pub block_number: u64,
+        /// The index of the event in the block.
+        pub log_index: usize,
+        pub uid: domain::OrderUid,
+        pub owner: eth::Address,
+        pub signed: bool,
+    }
+}
+
+impl Events {
+    pub fn new(persistence: Arc<infra::Persistence>) -> Self {
+        Self { persistence }
+    }
+
+    pub async fn latest_block(&self) -> Result<u64, Error> {
+        self.persistence
+            .latest_settlement_event_block()
+            .await
+            .map_err(Error::Persistence)
+    }
+
+    pub async fn append(
+        &self,
+        events: Vec<boundary::Event<boundary::events::GPv2Contract>>,
+    ) -> Result<(), Error> {
+        let mut tx = self.persistence.begin().await?;
+        self.insert(&mut tx, events).await?;
+        tx.commit().await.map_err(Error::Persistence)
+    }
+
+    pub async fn replace(
+        &self,
+        events: Vec<boundary::Event<boundary::events::GPv2Contract>>,
+        range: RangeInclusive<u64>,
+    ) -> Result<(), Error> {
+        let mut tx = self.persistence.begin().await?;
+        self.delete(&mut tx, range).await?;
+        self.insert(&mut tx, events).await?;
+        tx.commit().await.map_err(Error::Persistence)
+    }
+
+    async fn delete(&self, tx: &mut Transaction, range: RangeInclusive<u64>) -> Result<(), Error> {
```

Comment:
This function should only take a starting block if it  deletes everything starting from there anyway.
Otherwise some caller might think they can delete a range of blocks that doesn't end at the current block.

Reply:
I have to double check the EventHandling logic, but I believe we are currently incorrectly removing everything (instead of just the range). I didn't want to change the logic here, but you are right that the domain should then also take only the `from_block` arg